src/backend/metal/src/window.rs

```diff
@@ -39,84 +39,144 @@ impl Drop for SurfaceInner {
+unsafe impl Send for Frame {}
+unsafe impl Sync for Frame {}
+
+impl Drop for Frame {
     fn drop(&mut self) {
-        for maybe in self.frames.drain(..) {
-            if let Some((drawable, _)) = maybe {
-                unsafe {
-                    msg_send![drawable, release];
-                }
-            }
-        }
+        info!("dropping Frame");
     }
 }
 
 pub struct Swapchain {
-    inner: Arc<RwLock<SwapchainInner>>,
+    frames: Arc<Vec<Frame>>,
     surface: Arc<SurfaceInner>,
-    _size_pixels: (u64, u64),
+    size_pixels: (image::Size, image::Size),
+    last_frame: usize,
+    image_ready_callbacks: Vec<Arc<Mutex<Option<SwapchainImage>>>>,
 }
 
-unsafe impl Send for Swapchain {}
-unsafe impl Sync for Swapchain {}
+impl Drop for Swapchain {
+    fn drop(&mut self) {
+        info!("dropping Swapchain");
+        for ir in self.image_ready_callbacks.drain(..) {
+            if ir.lock().unwrap().take().is_some() {
+                debug!("\twith a callback");
+            }
+        }
+    }
+}
 
 impl Swapchain {
-    pub(crate) fn present(&self, index: hal::FrameImage) {
-        let (drawable, _) = self.inner
-            .write()
-            .unwrap()
+    /// Returns the drawable for the specified swapchain image index,
+    /// marks the index as free for future use.
+    pub(crate) fn take_drawable(&self, index: hal::SwapImageIndex) -> metal::Drawable {
+        let mut frame = self
             .frames[index as usize]
+            .inner
+            .lock()
+            .unwrap();
+        assert!(!frame.available);
+        frame.available = true;
+        frame.drawable
             .take()
-            .expect("Frame is not ready to present!");
-        unsafe {
-            msg_send![drawable, present];
-            msg_send![drawable, release];
+            .expect("Drawable has not been acquired!")
+    }
+
+    fn signal_sync(&self, sync: hal::FrameSync<Backend>) {
+        match sync {
+            hal::FrameSync::Semaphore(semaphore) => {
+                if let Some(ref system) = semaphore.system {
+                    system.signal();
+                }
+            }
+            hal::FrameSync::Fence(fence) => {
+                *fence.mutex.lock().unwrap() = true;
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct SwapchainImage {
+    frames: Arc<Vec<Frame>>,
+    surface: Arc<SurfaceInner>,
+    index: hal::SwapImageIndex,
+}
+
+impl SwapchainImage {
+    /// Waits until the specified swapchain index is available for rendering.
+    pub fn wait_until_ready(&self) {
+        // check the target frame first
+        {
+            let frame = self.frames[self.index as usize].inner.lock().unwrap();
+            assert!(!frame.available);
+            if frame.drawable.is_some() {
+                return
+            }
+        }
+        // wait for new frames to come until we meet the chosen one
+        while self.surface.next_frame(&self.frames).0 != self.index as usize {
```

Comment:
Is there a low-cost alternative to spinning here? I don't know how long these waits commonly last, but it seems like there's potential to run faster than presents occur sometimes